### ISSUE TITLE
[Research] Add Multi-Source Domain Adaptation for Cross-Subject EEG Emotion (MS-DCDA: 90.84% SEED)

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#113]
## Paper References

1. **"Multi-Source EEG Emotion Recognition via Dynamic Contrastive Domain Adaptation"** — Xiao et al., Biomedical Signal Processing and Control, 2024 (arXiv:2408.10235). MS-DCDA achieves **90.84% cross-subject on SEED** and **78.49% on SEED-IV**. Uses multi-branch contrastive neural network with DE features across 5 bands. Dynamically weighted domain alignment balances transferability and discriminability.

2. **"Hybrid transfer learning strategy for cross-subject EEG emotion recognition (DFF-Net)"** — Frontiers in Human Neuroscience, 2023. DFF-Net combines domain adaptation + few-shot fine-tuning. **93.37% on SEED**, **82.32% on SEED-IV**. Only needs ~5 labeled samples per class from target user.

3. **"Cross-subject EEG signals-based emotion recognition using contrastive learning (CSCL)"** — Scientific Reports, 2025. Cross-subject contrastive learning in hyperbolic space with region-aware learning. **97.70% on SEED**, 65.98% on FACED, 51.30% on MPED. Region-aware mechanism learns to focus on channel combinations for specific brain areas.

4. **"Dynamic domain adaptive EEG emotion recognition based on multi-source selection"** — Review of Scientific Instruments, 2025. Treats each subject as separate domain, filters similar source domains, extracts common+domain-specific features. **89.76% cross-subject on SEED**, 65.28% on SEED-IV.

5. **"Harnessing Few-Shot Learning for EEG signal classification: a survey"** — Frontiers in Human Neuroscience, 2024. Comprehensive survey of data augmentation, transfer learning, and self-supervised learning for few-shot EEG. Prototypical Networks with instance-attention achieves superior few-shot performance on DEAP/SEED.

## Key Contribution

These methods solve the biggest accuracy problem in the current system: cross-subject emotion recognition. The current feature-based heuristics achieve ~55-65% on new users. Domain adaptation techniques can push this to 75-90% by treating each user's data distribution as a separate domain and learning domain-invariant representations. The DFF-Net approach is especially practical — it needs only ~5 labeled samples per class from a new user for fine-tuning.

## How It Applies to NeuralDreamWorkshop

- **Immediate accuracy improvement**: Replace population-average thresholds with domain-adapted features
- **Personalization pipeline**: DFF-Net's few-shot fine-tuning integrates with existing BaselineCalibrator
- **Multi-user support**: MS-DCDA can leverage data from multiple users to improve model for new users
- **Bridge to per-user models**: Start with population model, adapt with 5 labeled samples, then continuously improve
- **Complements #40 (Contrastive Learning)**: CSCL provides the representation learning; MS-DCDA/DFF-Net provide the domain adaptation strategy

### New file: `ml/models/emotion_domain_adapter.py`

```python
import torch
import torch.nn as nn
import numpy as np
from typing import Dict, List, Optional

class DomainAdaptiveEmotionClassifier:
    """Cross-subject emotion recognition via domain adaptation.
    
    Implements DFF-Net approach:
    1. Pre-train feature extractor on multi-user DE features
    2. Domain-adapt to reduce distribution shift
    3. Few-shot fine-tune on 5 labeled samples per class from target user
    """
    
    def __init__(self):
        self.feature_extractor = None  # shared CNN/MLP backbone
        self.domain_discriminator = None  # adversarial domain classifier
        self.emotion_classifier = None  # task classifier
        self.user_prototypes = {}  # per-class prototypes for few-shot
        self.source_domains = []  # learned source user distributions
    
    def extract_de_features(self, eeg: np.ndarray, fs: float = 256.0) -> np.ndarray:
        """Extract Differential Entropy features across 5 bands x 4 channels.
        
        Returns 20-dim DE vector + 5 DASM + 5 RASM + 1 FAA = 31 features
        (matching the system's existing feature extraction pipeline).
        """
        from processing.eeg_processor import (
            differential_entropy, preprocess, compute_dasm_rasm,
            compute_frontal_asymmetry
        )
        
        features = []
        channels = eeg if eeg.ndim == 2 else eeg.reshape(1, -1)
        n_ch = min(channels.shape[0], 4)
        
        bands = {
            'delta': (1, 4), 'theta': (4, 8), 'alpha': (8, 13),
            'beta': (13, 30), 'gamma': (30, 45)
        }
        
        # DE per channel per band (20 features)
        for ch in range(n_ch):
            processed = preprocess(channels[ch], fs)
            for band_name, (low, high) in bands.items():
                de = differential_entropy(processed, fs)
                features.append(de.get(band_name, 0.0))
        
        # DASM + RASM (10 features) — already implemented
        if n_ch >= 4:
            dasm_rasm = compute_dasm_rasm(channels, fs)
            for band_name in bands:
                features.append(dasm_rasm.get(f'dasm_{band_name}', 0.0))
            for band_name in bands:
                features.append(dasm_rasm.get(f'rasm_{band_name}', 1.0))
        
        # FAA (1 feature)
        if n_ch >= 4:
            faa = compute_frontal_asymmetry(channels, fs, left_ch=1, right_ch=2)
            features.append(faa.get('frontal_asymmetry', 0.0))
        
        return np.array(features, dtype=np.float32)
    
    def adapt_to_user(self, labeled_samples: List[Dict], 
                      n_shots: int = 5) -> dict:
        """Few-shot adaptation with labeled user data.
        
        Args:
            labeled_samples: list of {eeg: ndarray, label: str}
            n_shots: samples per class (default 5 = 30 total for 6 classes)
        
        Strategy (DFF-Net):
        1. Extract DE features from labeled samples
        2. Compute per-class prototypes (mean embedding)
        3. Fine-tune classifier head on user prototypes
        """
        from collections import defaultdict
        class_features = defaultdict(list)
        
        for sample in labeled_samples:
            de = self.extract_de_features(sample['eeg'])
            class_features[sample['label']].append(de)
        
        # Compute prototypes
        self.user_prototypes = {}
        for label, feats in class_features.items():
            self.user_prototypes[label] = np.mean(feats, axis=0)
        
        return {
            'n_classes': len(self.user_prototypes),
            'samples_per_class': {k: len(v) for k, v in class_features.items()},
            'adapted': True
        }
    
    def predict(self, eeg: np.ndarray, fs: float = 256.0) -> dict:
        """Predict emotion using domain-adapted model.
        
        If adapted: use prototype matching (cosine similarity)
        If not adapted: fall back to existing emotion classifier
        """
        de_features = self.extract_de_features(eeg, fs)
        
        if self.user_prototypes:
            # Prototype-based classification (few-shot)
            similarities = {}
            for label, prototype in self.user_prototypes.items():
                cos_sim = np.dot(de_features, prototype) / (
                    np.linalg.norm(de_features) * np.linalg.norm(prototype) + 1e-10
                )
                similarities[label] = float(cos_sim)
            
            # Softmax over similarities
            max_sim = max(similarities.values())
            exp_sims = {k: np.exp((v - max_sim) * 5) for k, v in similarities.items()}
            total = sum(exp_sims.values())
            probabilities = {k: v / total for k, v in exp_sims.items()}
            
            predicted = max(probabilities, key=probabilities.get)
            
            return {
                'emotion': predicted,
                'probabilities': probabilities,
                'model_type': 'domain-adapted',
                'confidence': float(probabilities[predicted]),
                'adaptation_status': 'personalized'
            }
        else:
            return {
                'emotion': 'neutral',
                'probabilities': {},
                'model_type': 'unadapted',
                'confidence': 0.0,
                'adaptation_status': 'needs_calibration'
            }
```

### API endpoints in `ml/api/routes.py`:

```python
@router.post("/emotion/adapt")          # submit labeled samples for adaptation
@router.get("/emotion/adaptation-status")
@router.post("/emotion/predict-adapted")  # use adapted model
@router.post("/emotion/reset-adaptation")
```

### Integration with existing calibration flow:

Wire into `BaselineCalibrator` → after baseline collected, prompt user for 5 labeled emotional clips → call `adapt_to_user()` → subsequent predictions use adapted model.

## Expected Accuracy / Performance

| Method | SEED (3-class) | SEED-IV (4-class) | Muse 2 Expected |
|--------|---------------|-------------------|-----------------|
| Current heuristics | ~55-60% | ~45-50% | ~55-65% |
| MS-DCDA | 90.84% | 78.49% | 70-80% (4-ch gap) |
| DFF-Net (5-shot) | 93.37% | 82.32% | 75-85% (4-ch gap) |
| CSCL | 97.70% | — | 80-85% (4-ch gap) |

The 4-channel gap is ~10-15 points below full-channel results. With few-shot personalization, expect 75-85% on Muse 2 — a major improvement over current 55-65%.

## Complexity: HIGH

Requires implementing domain adaptation training loop (PyTorch), collecting/storing labeled user data, and managing per-user adapted models. The few-shot prototype approach (simpler part) can be implemented first. Full MS-DCDA with adversarial training is more complex but delivers better results.

## Muse 2 Specific Notes

- DE features from 4 channels produce 20-dim vector (vs 160-dim for 32 channels) — simpler model needed
- DASM/RASM features (already implemented) are critical for domain adaptation — they capture the asymmetry patterns that transfer across subjects
- Few-shot approach (5 samples/class) means user watches 30 short emotion clips (~5 min) for calibration
- The existing `online_learner.py` provides the adaptation infrastructure — wire it into this pipeline
- Start with DFF-Net's simpler fine-tuning approach before attempting full MS-DCDA adversarial training
